File: apps/runner/modules/data/experiment.py
from typing import Dict, Optional
from enum import Enum
from pydantic import BaseModel, root_validator, validator
from modules.data.configData import ConfigData

from modules.data.parameters import Parameter
from modules.data.types import DocumentId, EpochMilliseconds, UserId


class ExperimentType(Enum):
    UNKNOWN = "unknown"
    PYTHON = "python"
    JAVA = "java"
    C = "c"
    ZIP = "zip"


class ExperimentData(BaseModel):

    experimentType = ExperimentType.UNKNOWN
    experimentExecutable = ""
    expId: DocumentId
    creator: UserId
    creatorRole: str
    trialExtraFile: Optional[str]
    trialResult: str
    file = ""  #Will be set either by initializing or by app.py
    timeout: int
    keepLogs: bool
    scatter: bool
    scatterIndVar: Optional[str]
    scatterDepVar: Optional[str]
    dumbTextArea: str
    postProcess = False

    hyperparameters: Dict[str, Parameter]
    configs = {}  #Will be set Later

    startedAtEpochMillis: Optional[EpochMilliseconds]
    finishedAtEpochMillis: Optional[EpochMilliseconds]
    finished = False  # TODO replace with presence of finished timestamp?

    totalExperimentRuns: int = 0
    passes: int = 0
    fails: int = 0

    @validator('trialResult')
    @classmethod
    def check_trialResult(cls, v):
        if v == '':
            raise ValueError("Trial Result field cannot be empty")
        return v

    @validator('configs')
    @classmethod
    def check_configs(cls, v):
        for key, param in v.items():
            if not param.__class__ == ConfigData:
                raise ValueError(f'value {param} associated with {key} in configs is not a ConfigData object')
        return v

    @validator('timeout')
    @classmethod
    def check_timeout(cls, v):
        if v < 1:
            raise ValueError(f'value {v} is an invalid timeout, timeout must be greater than or equal to 1')
        return v

    @validator('hyperparameters')
    @classmethod
    def check_hyperparams(cls, v):
        for key, param in v.items():
            # For some reason, isinstance does not work here. Maybe it has to do with how pydantic validators work? - Rob
            if not param.__class__ in Parameter.__subclasses__():
                raise ValueError(f'value {param} associated with {key} in hyperparameters is not a Parameter')
            if key == '':
                raise ValueError('Key for parameter cannot be empty')
        return v

    @root_validator
    @classmethod
    def check_scatter_settings(cls, values):
        scatter, scatterIndVar, scatterDepVar = values.get('scatter'), values.get('scatterIndVar'), values.get('scatterDepVar')
        if scatter:
            if scatterIndVar == '' or scatterDepVar == '':
                raise ValueError("scatter is enabled, but scatterIndVar and/or scatterDepVar are absent")
        elif scatterIndVar != '' or scatterDepVar != '':
            raise ValueError("scatter is disabled, but scatterIndVar and/or scatterDepVar are present")
        return values

    def has_extra_files(self):
        return self.trialExtraFile != ''

    class Config:
        validate_assignment = True

File: apps/GladosBackend/Models/Experiment.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

public class Experiment
{
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public string? Id { get; set; }

    [BsonElement("hyperparameters")]
    public BsonDocument? Hyperparameters { get; set; }

    [BsonElement("name")]
    public string? Name { get; set; }

    [BsonElement("description")]
    public string? Description { get; set; }

    [BsonElement("trialExtraFile")]
    public string? TrialExtraFile { get; set; }

    [BsonElement("trialResult")]
    public string? TrialResult { get; set; }

    [BsonElement("verbose")]
    public bool Verbose { get; set; }

    [BsonElement("workers")]
    public int Workers { get; set; }

    [BsonElement("scatter")]
    public bool Scatter { get; set; }

    [BsonElement("dumbTextArea")]
    public string? DumbTextArea { get; set; }

    [BsonElement("scatterIndVar")]
    public string? ScatterIndVar { get; set; }

    [BsonElement("scatterDepVar")]
    public string? ScatterDepVar { get; set; }

    [BsonElement("timeout")]
    public int Timeout { get; set; }

    [BsonElement("keepLogs")]
    public bool KeepLogs { get; set; }

    [BsonElement("creator")]
    public string? Creator { get; set; }

    [BsonElement("created")]
    [BsonRepresentation(BsonType.Int64)]
    public long Created { get; set; } // Epoch timestamp (milliseconds)

    [BsonElement("finished")]
    public bool Finished { get; set; }

    [BsonElement("estimatedTotalTimeMinutes")]
    public double EstimatedTotalTimeMinutes { get; set; }

    [BsonElement("totalExperimentRuns")]
    public int TotalExperimentRuns { get; set; }

    [BsonElement("startedAtEpochMillis")]
    [BsonRepresentation(BsonType.Int64)]
    public long StartedAtEpochMillis { get; set; }

    [BsonElement("passes")]
    public int Passes { get; set; }
}